README.md
![VTU Results](https://raw.github.com/maheshkkumar/VTUResults/master/vtu.jpg)

# VTUResults
Python Package to fetch [VTU](http://results.vtu.ac.in) Results.

| Build Status | Dependency Status | Version | License | Downloads |
| ------------ | ------------- | ------- | -------------------|--------------- |
| [![Build Status](https://travis-ci.org/maheshkkumar/VTUResults.svg?branch=master)](https://travis-ci.org/maheshkkumar/VTUResults) | [![Dependency Status](https://gemnasium.com/maheshkkumar/VTUResults.svg)](https://gemnasium.com/maheshkkumar/VTUResults) | [![Foo](https://img.shields.io/badge/version-v2.0.1-blue.svg)](https://pypi.python.org/pypi/VTUResults/2.0.1) | [![GitHub license](https://img.shields.io/badge/license-GPLv2-blue.svg)](https://raw.githubusercontent.com/maheshkkumar/VTUResults/master/LICENSE) |[![Foo](https://img.shields.io/badge/downloads-6k%2Fmonth-blue.svg)](https://pypi.python.org/pypi/VTUResults/2.0.1) |





Features
========

- Compatible with Python 2 (2.7+)
- Get Result details for any USN
- All the results will be exported to a .txt file in the current working directory
- In `get_entire_result` method, `rank_list.txt` and `result_class.txt` will be created, the former file will contain the rank list of all the students and the latter file will contain the entire result data of all the students.
- In `get_group_usn` method, `results.txt` will be created in the current working directory, this file will contain all the result data of your chosen USN.
- In `get_usn` method, `result.txt` will be created and it will contain the result of a single USN.

Installation
========

    $ sudo pip install VTUResults

Usage
========

    from vr import VR

    result =  VR()

    # print the result of the USN sent as parameter to get_usn method.
    result.get_usn('1XX12XX100')
    result.get_group_usn()
    result.get_entire_result()
    
    Every result will be exported to a text file in your current working directory.
    
    
Package Reference
========

## Class: `VR`

### Get Result from VTUResults

#### `get_usn(usn)`

**Parameters:**

| Name | Type | Required | Description | Default |
| ---- | ---- | -------- | ----------- | ------- |
| `usn` | string | Yes | Returns the Result of the USN, passed as the parameter to get_usn function | `None` |

#### `get_group_usn()`

**Parameters:**

| Name | Type | Required | Description | Default |
| ---- | ---- | -------- | ----------- | ------- |
| `None` | `None`  |  `No` | Returns the Result of the total number of USN entered through terminal| `None` |

#### `get_entire_result()`

**Parameters:**

| Name | Type | Required | Description | Default |
| ---- | ---- | -------- | ----------- | ------- |
| `None` | `None`  | `No`  | Returns the Result of the entire department| `None` |



Example
========

    from vr import VR
    result =  VR()
    result.get_usn('1XX12XX100')
    result.get_group_usn()
    result.get_entire_result()
    
    # get_entire_result method will create a rank_file.txt in your current working directory, this file contains the rank list of all the students with their respective total marks.



Python Dependencies
========
##### [requests](http://docs.python-requests.org/en/latest/)

    $ pip install requests

Requests is a Python HTTP library, released under the Apache2 License. The goal of the project is to make HTTP requests simpler and more human-friendly.

Requests takes all of the work out of Python HTTP/1.1 — making your integration with web services seamless. There’s no need to manually add query strings to your URLs, or to form-encode your POST data. Keep-alive and HTTP connection pooling are 100% automatic, powered by urllib3, which is embedded within Requests

##### [beautifulsoup4](http://www.crummy.com/software/BeautifulSoup/)

    $ pip install beautifulsoup4

Beautiful Soup is a Python library designed for quick turnaround projects like screen-scraping.



CHANGELOG.md
## v2.0.1
- Fix: Bug Fixes and versioning the package

## v2.0.0
- Fix: Bug Fixes and versioning the package

## v1.9.0
- Fix: Bug Fixes and versioning the package

## v1.8.0
- Fix: Bug Fixes and versioning the package

## v1.7.0
- Fix: Bug Fixes and versioning the package

## v1.6.0
- Fix: Versioning the package

## v1.5.0
- Fix: Versioning the package

## v1.4.0
- Fix: Versioning the package

## v1.3.0
- Fix: Added Rank List to get_entire_result method

## v1.2.0
- Fix: Added New Methods

## v1.1.0
- Fix: Improved Exception Handling

## v1.0.6
- Fix: Added Exception Handling

## v1.0.5
- Fix: Versioning on PyPi

## v1.0.4
- Fix: Versioning on PyPi

## v1.0.3
- Fix: Changed from lxml to BeautifulSoup4

## v1.0.2
- Fix: Versioning on PyPi

## v1.0.1

- Fix: Versioning on PyPi


vtu_result.py
#!/usr/bin/python   
# -*- coding: utf-8 -*-
#######################################  
# Created by Mahesh Kumar K           #
# Python Script to fetch VTU Results  #
#######################################

import requests
from lxml import html

base_url = 'http://results.vtu.ac.in/vitavi.php?rid='
usn = []

def rolno():
    choice = input("Press 1 for single result and 2 for multiple results : ")
    if choice==1:
        s_usn = raw_input('Enter your usn : ')
        if len(s_usn) == 10:
            complete_url = base_url+s_usn+'&submit=SUBMIT'
            response = requests.get(complete_url)
            text = html.fromstring(response.text)
            print_marks(text)    
        else :
            print "Invalid USN"
    else:
       num = input("Enter the number of results, that you want to get : ")
       for i in range(0, num):
           s_usn = raw_input('Enter your usn : ')
           usn.append(s_usn)
           print usn
       for i in range(0,num):
           if len(usn[i]) == 10:
               complete_url = base_url+usn[i]+'&submit=SUBMIT'
               response = requests.get(complete_url)
               text = html.fromstring(response.text)
               print_marks(text)    
           else :
               print "Invalid USN"

def print_marks(html_data):
    subjects = html_data.xpath('//i/text()')
    marks = html_data.xpath('//td[@width=60][@align="center"]/text()')
    student_details = html_data.xpath('//b/text()')
    t_marks = html_data.xpath('//td/text()')
    total_marks = int(t_marks[61])
    student_name = student_details[0]
    semester = int(student_details[2])
    result = student_details[3].encode('ascii','ignore')
    print "******************************************************"
    print "Name : "+student_name
    print "Semester: ",semester
    print "Marks format is : Subject Name / Internal / External / Total"
    if subjects:
        internal = marks[5::3]
        external = marks[4::3]
        total = marks[6::3]     
        for i,j,k,l in zip(subjects, internal, external, total):
            print i,j,k,l
        print "Total Marks = ",total_marks
        if semester == 8:
            print "Average = ",round(float(total_marks * 100)/750, 2)
        elif semester == 1 or semester == 2:
            print "Average = ",round(float(total_marks * 100)/775, 2)
        else:
            print "Average = ",round(float(total_marks * 100)/900, 2)
        print "Result = "+result[8:]   
        print "******************************************************"
    else:
        print "Enter a Valid USN"
    
if __name__ == '__main__':
    rolno()
    
    


vtu_results/vtu_results.py
#!/usr/bin/env python   
# -*- coding: utf-8 -*-
#######################################  
# Created by Mahesh Kumar K           #
# Python Script to fetch VTU Results  #
#######################################
import requests
from lxml import html

from constants import BASE_URL
from utils import get_result

class VR(object):
  def __init__(self):
    pass

  def get_url(self, usn):
    self.usn = usn
    if len(usn) == 10:
      html = get_result(usn)
      subjects = html.xpath('//i/text()')
      marks = html.xpath('//td[@width=60][@align="center"]/text()')
      student_details = html.xpath('//b/text()')
      t_marks = html.xpath('//td/text()')
      total_marks = int(t_marks[61])
      student_name = student_details[0]
      semester = int(student_details[2])
      result = student_details[3].encode('ascii','ignore')

      print "Name : "+student_name
      print "Semester: ",semester
      print "Marks format is : Subject Name / Internal / External / Total"
      if  subjects:
          internal = marks[5::3]
          external = marks[4::3]
          total = marks[6::3]     
          for i,j,k,l in zip(subjects, internal, external, total):
              print i,j,k,l
          print "Total Marks = ",total_marks
          if semester == 8:
              print "Average = ",round(float(total_marks * 100)/750, 2)
          elif semester == 1 or semester == 2:
              print "Average = ",round(float(total_marks * 100)/775, 2)
          else:
              print "Average = ",round(float(total_marks * 100)/900, 2)
          print "Result = "+result[8:]
          print "complete results obtained"

    else:
      return "Invalid USN"



'''
import requests
from lxml import html

base_url = 'http://results.vtu.ac.in/vitavi.php?rid='
usn = []

def rolno():
    choice = input("Press 1 for single result and 2 for multiple results : ")
    if choice==1:
        s_usn = raw_input('Enter your usn : ')
        if len(s_usn) == 10:
            complete_url = base_url+s_usn+'&submit=SUBMIT'
            response = requests.get(complete_url)
            text = html.fromstring(response.text)
            print_marks(text)    
        else :
            print "Invalid USN"
    else:
       num = input("Enter the number of results, that you want to get : ")
       for i in range(0, num):
           s_usn = raw_input('Enter your usn : ')
           usn.append(s_usn)
           print usn
       for i in range(0,num):
           if len(usn[i]) == 10:
               complete_url = base_url+usn[i]+'&submit=SUBMIT'
               response = requests.get(complete_url)
               text = html.fromstring(response.text)
               print_marks(text)    
           else :
               print "Invalid USN"

def print_marks(html_data):
    subjects = html_data.xpath('//i/text()')
    marks = html_data.xpath('//td[@width=60][@align="center"]/text()')
    student_details = html_data.xpath('//b/text()')
    t_marks = html_data.xpath('//td/text()')
    total_marks = int(t_marks[61])
    student_name = student_details[0]
    semester = int(student_details[2])
    result = student_details[3].encode('ascii','ignore')
    print "******************************************************"
    print "Name : "+student_name
    print "Semester: ",semester
    print "Marks format is : Subject Name / Internal / External / Total"
    if subjects:
        internal = marks[5::3]
        external = marks[4::3]
        total = marks[6::3]     
        for i,j,k,l in zip(subjects, internal, external, total):
            print i,j,k,l
        print "Total Marks = ",total_marks
        if semester == 8:
            print "Average = ",round(float(total_marks * 100)/750, 2)
        elif semester == 1 or semester == 2:
            print "Average = ",round(float(total_marks * 100)/775, 2)
        else:
            print "Average = ",round(float(total_marks * 100)/900, 2)
        print "Result = "+result[8:]   
        print "******************************************************"
    else:
        print "Enter a Valid USN"
    
if __name__ == '__main__':
    rolno()
     
'''